### PR TITLE
Function trigger

### PR DESCRIPTION
**Pull request by resdenia**
-Exclude function from cloud storage
-Change trigger from HTTP to Pubsub trigger

### CODE REVIEW

File: run.sh [hunk 1]
```diff
@@ -206,6 +206,17 @@ function run_cloud_build(){
     # Run project
     cmd_create_cloud_build="$(curl -X POST -T config.json -H "Authorization: Bearer $access_token" https://cloudbuild.googleapis.com/v1/projects/$project_id/builds)"
 
+
+	# Create Function with using local files
+    function_name_sufix="${function_name}_func_logzio"
+	topic_prefix="$function_name-pubsub-topic-logs-to-logzio"
```
Comment (Doron-Bargo) [inline run.sh]: Please fix indentation

File: config.json [hunk 2]
```diff
@@ -83,38 +83,17 @@
...
-                "gcloud alpha functions add-iam-policy-binding $_FUNCTION_NAME --region=$_REGION --member=allUsers --role=roles/cloudfunctions.invoker\n"
-            ]
         }
     ],
     "substitutions": {
         "_PUBSUB_TOPIC_NAME": "pubsub-topic-logs-to-logzio",
         "_PUBSUB_SUBSCRIPTION_NAME": "pubsub-subscription-logs-to-logzio",
         "_SINK_NAME": "sink-logs-to-logzio",
-        "_FUNCTION_NAME": "func_logzio",
+        "_FUNCTION_NAME": "logzio_func_logzio",
```
Comment (Doron-Bargo) [inline config.json]: why this name?
If it for logs why not
logzio_logs_function
Or something similar

File: cloud_function_go/function.go [hunk 3]
```diff
@@ -70,49 +80,75 @@ func (l *logzioConfig) validateAndPopulateArguments(r *http.Request) {
...
 }
 
+func updateFields(rawDecodedText []byte) ([]byte, error) {
+	var m map[string]interface{}
+	err := json.Unmarshal(rawDecodedText, &m)
+	if err != nil {
+		fmt.Printf("Can't parse a json: %s", err)
+		return nil, err
+	}
+	val, ok := m["textPayload"]
```
Comment (Doron-Bargo) [inline cloud_function_go/function.go]: please use const parameters for "textPayload" and "severity" and

File: config.json [hunk 4]
```diff
@@ -1,124 +1,98 @@
...
     }
+  ],
+  "substitutions": {
+    "_PUBSUB_TOPIC_NAME": "finaltest-pubsub-topic-logs-to-logzio",
+    "_PUBSUB_SUBSCRIPTION_NAME": "finaltest-pubsub-subscription-logs-to-logzio",
+    "_SINK_NAME": "finaltest-sink-logs-to-logzio",
+    "_FUNCTION_NAME": "finaltest_func_logzio",
+    "_REGION": "us-central1",
+    "_LOGZIO_LISTENER": "listener.logz.io",
+    "_LOGZIO_TOKEN": "vfFDZjscgjhUmUmhafqXaaTKBYNCZwpL",
```
Comment (Doron-Bargo) [inline config.json]: Why you have private token inside public repo?

File: run.sh [hunk 5]
```diff
@@ -174,24 +174,59 @@ function populate_data_to_json (){
...
+function get_project_id(){
+    echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")] Getting Google Project ID..."
+
+    gcloud config get-value project
+    if [[ $? -ne 0 ]]; then
+        echo -e "[ERROR] [$(date +"%Y-%m-%d %H:%M:%S")] Failed to get user project id  ..."
+        exit 1
+    else
+        project_id="$(gcloud config get-value project)"
+        echo -e "[INFO] [$(date +"%Y-%m-%d %H:%M:%S")] Project ID=$project_id where will launch integration."
```
Comment (Doron-Bargo) [inline run.sh]: Integration will be launch in Project ID=...

**Review by Doron-Bargo — CHANGES_REQUESTED** [review]
Please check my comments

File: cloud_function_go/function.go [hunk 6]
```diff
@@ -70,49 +83,75 @@ func (l *logzioConfig) validateAndPopulateArguments(r *http.Request) {
...
+		return err
+	}
+	return nil
+}
+
 func doRequest(rawDecodedText []byte, url string) {
+
+	err := updateFields(&rawDecodedText)
+	if err != nil {
+		fmt.Printf("Can't to parse json object: %s", err)
```
Comment (Doron-Bargo) [inline cloud_function_go/function.go]: Can't update fields

File: cloud_function_go/function.go [hunk 7]
```diff
@@ -70,49 +83,75 @@ func (l *logzioConfig) validateAndPopulateArguments(r *http.Request) {
...
+	value, okey := m[severity]
+	// If the key severity exists
+	if okey {
+		delete(m, severity)
+		m["log_level"] = value
+	}
+
+	*rawDecodedText, err = json.Marshal(m)
+	if err != nil {
+		fmt.Printf("Can't parse to bytes: %s", err)
```
Comment (Doron-Bargo) [inline cloud_function_go/function.go]: Can't parse json

File: cloud_function_go/function.go [hunk 8]
```diff
@@ -70,49 +83,75 @@ func (l *logzioConfig) validateAndPopulateArguments(r *http.Request) {
...
+
 func doRequest(rawDecodedText []byte, url string) {
+
+	err := updateFields(&rawDecodedText)
+	if err != nil {
+		fmt.Printf("Can't to parse json object: %s", err)
+		return
+	}
 	if binary.Size(rawDecodedText) > maxSize {
 		fmt.Printf("The request body size is larger than %d KB.", maxSize)
```
Comment (Doron-Bargo) [inline cloud_function_go/function.go]: The request body size is larger than %d KB. We will cut the ...

File: cloud_function_go/function.go [hunk 9]
```diff
@@ -70,49 +83,75 @@ func (l *logzioConfig) validateAndPopulateArguments(r *http.Request) {
...
-		req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(rawDecodedText))
+		req, err := http.NewRequest("POST", url, &compressedBuf)
 		if err != nil {
-			fmt.Printf("Connection was failed: %w", err)
+			fmt.Printf("Connection was failed: %s", err)
 			return
 		}
 		req.Header.Add("Content-Encoding", "gzip")
+		req.Header.Add("Content-Type", "text/plain")
+		req.Header.Add("logzio-shipper", "logzio-go/v1.0.0")
```
Comment (Doron-Bargo) [inline cloud_function_go/function.go]: why you use this header ? its not logzio-go shipper

File: cloud_function_go/function.go [hunk 10]
```diff
@@ -123,40 +162,21 @@ func doRequest(rawDecodedText []byte, url string) {
...
-		}
+func LogzioHandler(ctx context.Context, m PubSubMessage) error {
 
-		rawDecodedText, err := base64.StdEncoding.DecodeString(d.Message.Data)
-		if err != nil {
-			fmt.Printf("Message log can't be parsed: %w", err)
-			return
-		}
+	logzioConfig := logzioConfig{}
+	logzioConfig.validation = true
```
Comment (Doron-Bargo) [inline cloud_function_go/function.go]: what is this part?
Why you need to create this parameter and then call for the if
What is the other use case that this parameter will have false value?

File: config.json (hunk 2 above)
Comment (Doron-Bargo) [reply config.json]: The name is still not good

File: run.sh [hunk 11]
```diff
@@ -174,24 +174,59 @@ function populate_data_to_json (){
     echo "${contents}" > config.json
     contents="$(jq --arg sink_prefix "${function_name}" '.substitutions._SINK_NAME = $sink_prefix+"-sink-logs-to-logzio"' config.json)"
     echo "${contents}" > config.json
-    if [[ ! -z "$resource_type" ]]; then
-    contents="$(jq --arg resource_type "${resource_type}" '.substitutions._FILTER_LOG = $resource_type' config.json)"
+    if [[ ! -z "$resource_list" ]]; then
+    contents="$(jq --arg resource_list "${resource_list}" '.substitutions._FILTER_LOG = $resource_list' config.json)"
```
Comment (Doron-Bargo) [inline run.sh]: Please fix indentation

**Review by Doron-Bargo — CHANGES_REQUESTED** [review]
Please check my comments

File: cloud_function_go/function.go (hunk 10 above)
Comment (resdenia) [reply cloud_function_go/function.go]: I moved parameter to initialization of logzioConfig. And added to else statement to the if logic.

File: config.json [hunk 12]
```diff
@@ -83,39 +75,18 @@
...
-                "gcloud alpha functions add-iam-policy-binding $_FUNCTION_NAME --region=$_REGION --member=allUsers --role=roles/cloudfunctions.invoker\n"
-            ]
         }
     ],
     "substitutions": {
         "_PUBSUB_TOPIC_NAME": "pubsub-topic-logs-to-logzio",
         "_PUBSUB_SUBSCRIPTION_NAME": "pubsub-subscription-logs-to-logzio",
         "_SINK_NAME": "sink-logs-to-logzio",
-        "_FUNCTION_NAME": "func_logzio",
+        "_FUNCTION_NAME": "dump_name_gcp_fucntion_name_func_logzio",
```
Comment (Doron-Bargo) [inline config.json]: you are not using this parameter in your code

**Review by Doron-Bargo — CHANGES_REQUESTED** [review]
please check my comment

**Review by Doron-Bargo — APPROVED** [review]
lgtm

**Pull request merged**